tapgarden/planter.go

```diff
@@ -797,32 +855,140 @@ func (c *ChainPlanter) finalizeBatch(params FinalizeParams) (*BatchCaretaker,
 	})
 
 	if err != nil {
-		return nil, fmt.Errorf("unable to store tapscript "+
-			"tree for minting batch: %w", err)
+		return fmt.Errorf("unable to store tapscript tree for minting "+
+			"batch: %w", err)
 	}
 
-	if rootHash != nil {
-		ctx, cancel = c.WithCtxQuit()
-		defer cancel()
-		err = c.cfg.Log.CommitBatchTapSibling(
-			ctx, c.pendingBatch.BatchKey.PubKey, rootHash,
+	// Update the batch by adding the sibling root hash and genesis TX.
+	updateBatch := func(batch *MintingBatch) error {
+		// Add the batch sibling root hash if present.
+		if rootHash != nil {
+			batch.tapSibling = rootHash
+		}
+
+		// Fund the batch with the specified fee rate.
+		batchKey := asset.ToSerialized(batch.BatchKey.PubKey)
+		batchTX, err := c.fundGenesisPsbt(ctx, batchKey, feeRate)
+		if err != nil {
+			return fmt.Errorf("unable to fund minting PSBT for "+
+				"batch: %x %w", batchKey[:], err)
+		}
+
+		batch.GenesisPacket = batchTX
+
+		return nil
+	}
+
+	switch {
+	// If we don't have a batch, we'll create an empty batch before funding
+	// and writing to disk.
+	case c.pendingBatch == nil:
+		newBatch, err := c.newBatch()
+		if err != nil {
+			return fmt.Errorf("unable to create new batch: %w", err)
+		}
+
+		err = updateBatch(newBatch)
+		if err != nil {
+			return err
+		}
+
+		// Now that we're done populating parts of the batch, write it
+		// to disk.
+		err = c.cfg.Log.CommitMintingBatch(ctx, newBatch)
+		if err != nil {
+			return err
+		}
+
+		c.pendingBatch = newBatch
+
+	// If we already have a batch, we need to attach the optional sibling
+	// root hash and fund the batch.
+	case c.pendingBatch != nil:
+		err = updateBatch(c.pendingBatch)
+		if err != nil {
+			return err
+		}
+
+		// Write the associated sibling root hash and TX to disk.
+		if c.pendingBatch.tapSibling != nil {
+			err = c.cfg.Log.CommitBatchTapSibling(
+				ctx, c.pendingBatch.BatchKey.PubKey, rootHash,
+			)
+			if err != nil {
+				return fmt.Errorf("unable to commit tapscript "+
+					"sibling for minting batch %w", err)
+			}
+		}
+
+		err = c.cfg.Log.CommitBatchTx(
+			ctx, c.pendingBatch.BatchKey.PubKey,
+			c.pendingBatch.GenesisPacket,
 		)
 		if err != nil {
-			return nil, fmt.Errorf("unable to commit tapscript "+
-				"sibling for minting batch %w", err)
+			return err
 		}
 	}
 
-	c.pendingBatch.tapSibling = rootHash
+	return nil
+}
+
+func (c *ChainPlanter) sealBatch(params SealParams) error {
+	return nil
+}
+
+// finalizeBatch creates a new caretaker for the batch and starts it.
+func (c *ChainPlanter) finalizeBatch(params FinalizeParams) (*BatchCaretaker,
+	error) {
+
+	var (
+		feeRate *chainfee.SatPerKWeight
+		err     error
+	)
+
+	// Process the finalize parameters.
+	params.FeeRate.WhenSome(func(fr chainfee.SatPerKWeight) {
+		feeRate = &fr
+	})
+
+	ctx, cancel := c.WithCtxQuit()
+	defer cancel()
+
+	params.SiblingTapTree.WhenSome(func(tn asset.TapscriptTreeNodes) {
```

Comment:
Can write alternatively as: 
```go
err := fn.MapOptionZ(params.SiblingTapTree, func(tn asset.TapscriptTreeNodes) error {
       return c.cfg.TreeStore.StoreTapscriptTree(ctx, tn)
})
```

Reply:
Since we're also returning the `rootHash` I think this doesn't work out.